sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -3465,6 +3472,14 @@ object functions {
 
   /**
    * Returns an array of elements for which a predicate holds in a given array.
+   * {{{
+   *   df.select(filter(col("s"), (x, i) => i % 2 === 0))
+   *   df.selectExpr("filter(col, (x, i) -> i % 2 == 0)")
+   * }}}
+   *
+   * @param column: the input array column
+   * @param f: (col, index) => predicate, the boolean predicate to filter the input column
+   *           given the index. Indices start at 0.
```

Comment:
is this consistent within Spark that the indices parameter starts with 0 in higher-order functions? @ueshin @HyukjinKwon

Reply:
`ArrayTransform` with index argument starts with `0`.
We might need to change it from `1`?

Reply:
what's the behavior of presto?

Reply:
Actually presto's `transform` or `filter` don't take index argument.

I remember we had a discussion about the index argument in the PR for `zip_with_index` (https://github.com/apache/spark/pull/21121#issuecomment-384885012).
And for `filter`, it was done later separately, but seems like the similar context https://issues.apache.org/jira/browse/SPARK-28962.

Reply:
It's too late to change now, let's keep using 0.